src/httpserver.cpp

```diff
@@ -146,6 +149,10 @@ static GlobalMutex g_httppathhandlers_mutex;
 static std::vector<HTTPPathHandler> pathHandlers GUARDED_BY(g_httppathhandlers_mutex);
 //! Bound listening sockets
 static std::vector<evhttp_bound_socket *> boundSockets;
+//! Track active requests
+static GlobalMutex g_requests_mutex;
+static std::condition_variable g_requests_cv;
+static std::unordered_set<evhttp_request*> g_requests GUARDED_BY(g_requests_mutex);
```

Comment:
I think this can also be a simple atomic counter which doesn't require a lock?

<details>
<summary>git diff on 6d3a848fc</summary>

```diff
diff --git a/src/httpserver.cpp b/src/httpserver.cpp
index 346c6d5cb..a058a9c49 100644
--- a/src/httpserver.cpp
+++ b/src/httpserver.cpp
@@ -152,7 +152,7 @@ static std::vector<evhttp_bound_socket *> boundSockets;
 //! Track active requests
 static GlobalMutex g_requests_mutex;
 static std::condition_variable g_requests_cv;
-static std::unordered_set<evhttp_request*> g_requests GUARDED_BY(g_requests_mutex);
+static std::atomic<size_t> g_requests_count;
 
 /** Check if a network address is allowed to access the HTTP server */
 static bool ClientAllowed(const CNetAddr& netaddr)
@@ -216,11 +216,9 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
 {
     // Track requests and notify when a request is completed.
     {
-        WITH_LOCK(g_requests_mutex, g_requests.insert(req));
-        g_requests_cv.notify_all();
+        ++g_requests_count;
         evhttp_request_set_on_complete_cb(req, [](struct evhttp_request* req, void*) {
-            auto n = WITH_LOCK(g_requests_mutex, return g_requests.erase(req));
-            assert(n == 1);
+            --g_requests_count;
             g_requests_cv.notify_all();
         }, nullptr);
     }
@@ -480,10 +478,10 @@ void StopHTTPServer()
     {
         WAIT_LOCK(g_requests_mutex, lock);
         g_requests_cv.wait(lock, []() EXCLUSIVE_LOCKS_REQUIRED(g_requests_mutex) {
-                if (!g_requests.empty()) {
-                    LogPrint(BCLog::HTTP, "Waiting for %d requests to stop HTTP server\n", g_requests.size());
-                }
-                return g_requests.empty();
+            if (g_requests_count > 0) {
+                LogPrint(BCLog::HTTP, "Waiting for %d requests to stop HTTP server\n", g_requests_count);
+            }
+            return g_requests_count == 0;
         });
     }
     if (eventHTTP) {
```
</details>

(cc @hebasto - I think this is a similar but less "controversial" approach than what I proposed offline)

Reply:
Yes, for this change alone it would be possible with a count I think but not for the follow-up change in [#19434](https://github.com/bitcoin/bitcoin/pull/19434) where the explicit requests are removed in a callback.

Reply:
Okay, makes sense. I'm not sure how controversial #19434 is, need to take a closer look. Would it make sense to use the atomic counter in this PR and then use a set in the follow-up? In case we can't get that merged, that would leave master in a better state, and I think the code churn is manageable?

Reply:
You still need a lock, otherwise, how would you wait for the count to be zero?

Reply:
I'm not sure I understand your comment - my diff (https://github.com/bitcoin/bitcoin/pull/26742#discussion_r1063664153) does not involve removing the lock, although it reduces reliance on it since a counter can be encapsulated with a more efficient `std::atomic<size_t>` as opposed to set mutations which require acquiring a lock. If I'm misunderstanding you, could you please clarify?

Reply:
Ah neverming, we could use `std::atomic<T>::wait`

Reply:
Anyway, I think having a set of connections is required for upcoming changes (like #19434) but you are right that for this branch the count is enough. I still tend to prefer as is but I would ACK a version with the counter.

Reply:
I prefer to keep it as is as well. I don't think #19434 will be controversial because it's not conceptually different, just an improvement. And using the count is just a minor simplification, the code works the same either way. We have put stranger things into the code base, like unused functions for example, as long as there were follow-ups where they would be used and the these follow-ups were following the same concept that was already ACKed.

Reply:
Alright, it's not my ideal approach but let's be pragmatic and keep it as is. Can be marked as resolved for me.